hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDistributionSetInvalidationManagement.java

```diff
@@ -0,0 +1,183 @@
+        final String tenant = tenantAware.getCurrentTenant();
+
+        if (shouldRolloutsBeCanceled(distributionSetInvalidation.getCancelationType(),
+                distributionSetInvalidation.isCancelRollouts())) {
+            final String handlerId = JpaRolloutManagement.createRolloutLockKey(tenant);
+            final Lock lock = lockRegistry.obtain(handlerId);
+            try {
+                if (!lock.tryLock(repositoryProperties.getDsInvalidationLockTimeout(), TimeUnit.SECONDS)) {
+                    throw new StopRolloutException("Timeout while trying to invalidate distribution sets");
+                }
+                try {
+                    invalidateDistributionSetsInTransaction(distributionSetInvalidation, tenant);
+                } finally {
+                    lock.unlock();
+                }
+            } catch (final InterruptedException e) {
+                LOG.error("InterruptedException while invalidating distribution sets {}!",
+                        distributionSetInvalidation.getDistributionSetIds(), e);
+                Thread.currentThread().interrupt();
+            }
+        } else {
+            // no lock is needed as no rollout will be stopped
+            invalidateDistributionSetsInTransaction(distributionSetInvalidation, tenant);
+        }
+    }
+
+    private void invalidateDistributionSetsInTransaction(final DistributionSetInvalidation distributionSetInvalidation,
+            final String tenant) {
+        DeploymentHelper.runInNewTransaction(txManager, tenant + "-invalidateDS", status -> {
+            distributionSetInvalidation.getDistributionSetIds().forEach(setId -> invalidateDistributionSet(setId,
+                    distributionSetInvalidation.getCancelationType(), distributionSetInvalidation.isCancelRollouts()));
+            return 0;
+        });
+    }
+
+    private void invalidateDistributionSet(final long setId, final CancelationType cancelationType,
+            final boolean cancelRollouts) {
+        final DistributionSet set = distributionSetManagement.getValidAndComplete(setId);
+        distributionSetManagement.invalidate(set);
+        LOG.debug("Distribution set {} set to invalid", setId);
+
+        if (shouldRolloutsBeCanceled(cancelationType, cancelRollouts)) {
+            rolloutManagement.cancelRolloutsForDistributionSet(set);
+        }
+
+        if (cancelationType != CancelationType.NONE) {
+            deploymentManagement.cancelActionsForDistributionSet(cancelationType, set);
+        }
+
+        targetFilterQueryManagement.cancelAutoAssignmentForDistributionSet(setId);
+    }
+
+    private boolean shouldRolloutsBeCanceled(final CancelationType cancelationType, final boolean cancelRollouts) {
+        return cancelationType != CancelationType.NONE || cancelRollouts;
+    }
+
+    @Override
+    public DistributionSetInvalidationCount countEntitiesForInvalidation(
```

Comment:
please don't use repository classes directly but rather extend management classes with corresponding permissions for read (e.g. HAS_AUTH_ROLLOUT_MANAGEMENT_READ)